### PR TITLE
Smaract API testing version (for review)

### PR DESCRIPTION
**Pull request by amuskens**

### CODE REVIEW

File: src/odemis/driver/smarpod.py [hunk 1]
```diff
@@ -400,15 +647,15 @@ def IsReferenced(self):
...
+
     def GetMoveStatus(self):
         """
         Gets the move status from the controller.
         Returns:
             SmarPodDLL.SMARPOD_MOVING is returned if moving
             SmarPodDLL.SMARPOD_STOPPED when stopped
             SmarPodDLL.SMARPOD_HOLDING when holding between moves
-            SmarPodDLL.SMARPOD_CALIBRATING when calibrating
+            SmarPodDLL.SMARPOD_CALIBRATING wsmarpod.SmarAct_MC(**CONFIG)hen calibrating
```
Comment (tepals) [inline src/odemis/driver/smarpod.py]: copy paste error

File: src/odemis/driver/smarpod.py [hunk 2]
```diff
@@ -668,7 +915,7 @@ def _cancelCurrentMove(self, future):
         # The difficulty is to synchronise correctly when:
         #  * the task is just starting (not finished requesting axes to move)
         #  * the task is finishing (about to say that it finished successfully)
-        logging.debug("Canceling current move")
+        logging.debug("Canceling current smarpod.SmarAct_MC(**CONFIG)move")
```
Comment (tepals) [inline src/odemis/driver/smarpod.py]: copy paste error, you might want to check if there are more

File: src/odemis/driver/smarpod.py [hunk 3]
```diff
@@ -279,9 +279,256 @@ def __init__(self, error_code):
...
+    Convert a coordinate dictionary (str) -> (double) to a SA_MC Pose C struct
+        pos: Coordinate dictionary (str) -> (double) of axis name to value
+    returns: a Pose (C struct)
+    raises ValueError if an unsupported axis name is input
+    """
+    pose = SA_MC_Pose()
+
+    # Note: internally, the system uses metres and degrees for rotation
+    for an, v in pos.items():
+        if an == "x":
```
Comment (tepals) [inline src/odemis/driver/smarpod.py]: couldn't you rewrite this whole for loop to:
```
pose.x = pos["x"]
pose.y = pos["y"]
pose.z = pos["z"]
pose.rx = math.degrees(pos["rx"])
pose.ry = math.degrees(pos["ry"])
pose.rz = math.degrees(pos["rz"])
```

Comment (amuskens) [reply src/odemis/driver/smarpod.py]: I think I did it this way because it is possible to have a pos dict with not all of the axes i.e. pos = {'x': 0.5}

File: scripts/acquire_cryo.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return (int): value to return to the OS as program exit code
+    """
+    # Get components
+    # stage = model.getComponent(role="scan-stage")
+    #scanner = model.getComponent(role="laser-mirror")
+    """
+
+    # parse arguments
+    parser = argparse.ArgumentParser(description="Run an acquisition "
+                                     "with Cryo hardware for testing")
```
Comment (pieleric) [inline scripts/acquire_cryo.py]: I don't quite understand what does this script (better than just `odemis-cli --acquire se-detector ...`), but in any case it should be in a separate PR from the extension of the SmarAct driver.

File: src/odemis/driver/smarpod.py [hunk 5]
```diff
@@ -279,9 +279,256 @@ def __init__(self, error_code):
...
+        except Exception:
+            raise AttributeError("Failed to find %s" % (name,))
+        func.__name__ = name
+        func.errcheck = self.sp_errcheck
+        return func
+
+    @staticmethod
+    def sp_errcheck(result, func, args):
+        """
+        Analyse the retuhwModelrn value of a call and raise an exception in case of
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: `hwModel` got dropped in the middle of the string.

File: src/odemis/driver/mcs2.py [hunk 6]
```diff
@@ -0,0 +1,999 @@
...
+    def __init__(self, error_code):
+        self.errno = error_code
+        super(SA_CTLError, self).__init__("Error %d. %s" % (error_code, SA_CTLDLL.err_code.get(error_code, "")))
+
+
+class SmarAct_MCS2(model.Actuator):
+
+    def __init__(self, name, role, locator, ref_on_init=False,axes=None, **kwargs):
+        """
+        A driver for a SmarAct MCS2 Actuator.
```
Comment (pieleric) [inline src/odemis/driver/mcs2.py]: Please describe more here what is the "MCS2" actuator. I'd say it's "the standard SmarAct motion controller".

File: src/odemis/driver/smarpod.py [hunk 7]
```diff
@@ -870,3 +1117,568 @@ def Smarpod_GetDLLVersion(self, p_major, p_minor, p_update):
         minor.value = 2
         update = _deref(p_update, c_uint)
         update.value = 3
+
+
+class SmarAct_MC(model.Actuator):
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: Let's call this class "MCTriGlide".

File: src/odemis/driver/smarpod.py [hunk 8]
```diff
@@ -870,3 +1117,568 @@ def Smarpod_GetDLLVersion(self, p_major, p_minor, p_update):
...
+
+    def __init__(self, name, role, locator, ref_on_init=False, linear_speed=0.01,
+                 rotary_speed=1, axes=None, **kwargs):
+        """
+        A driver for a SmarAct SA_MC Actuator.
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: SA_MC -> MC. Please explain what the actuator is. That's the "Motion Controller" which formally was called "TriGlide", and is a 5DoF custom-designed stage.

File: src/odemis/driver/mcs2.py [hunk 9]
```diff
@@ -0,0 +1,999 @@
...
+    def _updateSpeed(self):
+        """
+        update the speeds
+        """
+        s = {}
+        try:
+            for axis_name, axis_channel in self._axis_map.items():
+                s[axis_name] = self.self.GetSpeed(axis_channel)
+                
+        except SA_CTLError as ex:
```
Comment (pieleric) [inline src/odemis/driver/mcs2.py]: Drop this try/except if you do nothing with the exception. (same for all the similar functions)

File: src/odemis/driver/mcs2.py [hunk 10]
```diff
@@ -0,0 +1,999 @@
...
+        self._checkMoveAbs(pos)
+        pos = self._applyInversion(pos)
+
+        f = self._createMoveFuture()
+        f = self._executor.submitf(f, self._doMoveAbs, f, pos)
+        return f
+
+    @isasync
+    def moveRel(self, shift):
+        self._checkMoveRel(shift)
```
Comment (pieleric) [inline src/odemis/driver/mcs2.py]: Make it like `moveAbs()`: immediately return if shift is empty.

File: src/odemis/driver/smarpod.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/driver/smarpod.py]: What happens to the axes where the pose is not defined? Do they move to a random position? Or back to 0? That's all sounds fishy...

File: src/odemis/driver/smarpod.py [hunk 11]
```diff
@@ -279,9 +279,256 @@ def __init__(self, error_code):
...
+        }
+
+    def __init__(self):
+        if os.name == "nt":
+            raise NotImplemented("Windows not yet supported")
+            # WinDLL.__init__(self, "libSA_MC.dll")  # TODO check it works
+            # atmcd64d.dll on 64 bits
+        else:
+            # Global so that its sub-libraries can access it
+            CDLL.__init__(self, "libsmaractmc.so.0.9.1", RTLD_GLOBAL)
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: It should be just `libsmaractmc.so.0` (which should be a link to the latest version of the library).

File: src/odemis/driver/smarpod.py [hunk 12]
```diff
@@ -279,9 +279,256 @@ def __init__(self, error_code):
...
+
+    def __init__(self, error_code):
+        self.errno = error_code
+        super(SA_MCError, self).__init__("Error %d. %s" % (error_code, SA_MCDLL.err_code.get(error_code, "")))
+
+
 class SmarPod(model.Actuator):
     
-    def __init__(self, name, role, locator, ref_on_init=False, actuator_speed=0.1,
+    def __init__(self, name, role, locator, daemon=None, ref_on_init=False, actuator_speed=0.1,
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: Don't get `daemon` here. It is normally a hidden data directly passed to the parent class (via kwargs). If you grab it here, the parent will never get it and be very sad.

File: src/odemis/driver/mcs2.py [hunk 13]
```diff
@@ -0,0 +1,999 @@
...
+
+        options = c_char_p(option_string.encode("ascii"))
+
+        self.core.SA_CTL_Open(byref(self._id), options)
+        logging.debug("Successfully connected to SA_CTL Controller ID %d", self._id.value)
+        model.Actuator.__init__(self, name, role, axes=axes_def, **kwargs)
+
+        # Add metadata
+        # TODO: Fix getting software version with a supported function
+        self._swVersion = 0  # self.GetSwVersion()
```
Comment (pieleric) [inline src/odemis/driver/mcs2.py]: Use `SA_CTL_GetFullVersionString()`?

File: src/odemis/driver/mcs2.py [hunk 14]
```diff
@@ -0,0 +1,999 @@
...
+
+        self._metadata[model.MD_SW_VERSION] = self._swVersion
+        logging.debug("Using SA_CTL library version %s", self._swVersion)
```
Comment (pieleric) [inline src/odemis/driver/mcs2.py]: Please also report the hwVersion, using SA_CTL_PKEY_FIRMWARE_VERSION.

File: src/odemis/driver/smarpod.py [hunk 15]
```diff
@@ -870,3 +1117,568 @@ def Smarpod_GetDLLVersion(self, p_major, p_minor, p_update):
...
+        self.core.SA_MC_Open(byref(self._id), options)
+        logging.debug("Successfully connected to SA_MC Controller ID %d", self._id.value)
+        model.Actuator.__init__(self, name, role, axes=axes_def, **kwargs)
+
+        # Add metadata
+        # TODO: Fix getting software version with a supported function
+        self._swVersion = 0  # self.GetSwVersion()
+
+        self._metadata[model.MD_SW_VERSION] = self._swVersion
+        logging.debug("Using SA_MC library version %s", self._swVersion)
```
Comment (pieleric) [inline src/odemis/driver/smarpod.py]: Let's ask SmarAct to get this info. For now, just disable entirely the code, so that the default value ("unknown") is used.

**Comment by pieleric** [comment]
BTW, to control the pivot point on the MCTriGlide class, the client shouldn't rely on directly calling the internal SetPivot()/GetPivot() methods. The class should provide a dedicated metadata, something like `MD_PIVOT_POS = {"x": ..., "y":..., "z":...)`. To support it, it should be mostly a matter of calling GetPivot() at init, and then override `updateMetadata()` to handle a change of value, by calling SetPivot().

File: src/odemis/driver/mcs2.py [hunk 16]
```diff
@@ -0,0 +1,999 @@
...
+        self._updatePosition()
+
+    def _updatePosition(self):
+        """
+        update the position VA
+        """
+        p = {}
+        try:
+            for axis_name, axis_channel in self._axis_map.items():
+                p[axis_name] = self.self.GetPosition(axis_channel)
```
[the remark this answers is not in the document]
Comment (amuskens) [reply src/odemis/driver/mcs2.py]: Haha... luckily this is fixed now...

File: src/odemis/driver/smarpod.py (hunk 7 above)
Comment (amuskens) [reply src/odemis/driver/smarpod.py]: Apparently, it's actually not a triglide. Any alternative ideas for the name?

File: src/odemis/driver/smarpod.py (hunk 3 above)
Comment (amuskens) [reply src/odemis/driver/smarpod.py]: I think it will move to 0 as that would be the value initialized. This could be problematic. Maybe this needs to be accounted for?

File: src/odemis/driver/test/smatpod_test.py [hunk 17]
```diff
@@ -42,8 +42,8 @@
         "role": "",
         "ref_on_init": True,
         "actuator_speed": 0.1,  # m/s
-        "locator": "usb:ix:0",
-        # "locator": "fake",
+        # "locator": "usb:ix:0",
+        "locator": "fake",
```
Comment (pieleric) [inline src/odemis/driver/test/smatpod_test.py]: Please use "usb:ix:0", and if `TEST_NOHW` is set then override "locator" to "fake".
